### PR TITLE
access to model

### PR DESCRIPTION
**Pull request by earonesty**
- access model
- create-from model
- use pytest fixtures for in-mem db
- row as-dict: keys are case-insensitive (same as sql), allow access as a "dict"
- translate errors to universal format

### CODE REVIEW

File: tests/test_notanorm.py [hunk 1]
```diff
@@ -1,182 +1,221 @@
...
+    db.create_model(model)
+    check = db.model()
+    assert check == model
 
-    def test_safedb_inmemorydb(self):
-        # test that in memory db's are relatively safe
 
-        db = SqliteDb(":memory:")
+def test_conn_retry(db):
+    db.query("create table foo (x integer)")
```
Comment (earonesty) [inline tests/test_notanorm.py:202]: this sneakily tests reconnect, should probably still assert that it fails when reconect is 0

File: tests/test_notanorm.py [hunk 2]
```diff
@@ -1,182 +1,221 @@
...
+def db_fixture(request, db_name):
+    yield request.getfixturevalue("db_" + db_name)
 
-    def tearDownClass():
-        log.setLevel(logging.INFO)
 
-    def test_db_basic(self):
-        fname = self.__fname()
+def pytest_generate_tests(metafunc):
+    global PYTEST_REG               # pylint: disable=global-statement
```
Comment (earonesty) [inline tests/test_notanorm.py:47]: some comment about what this does would be nice

File: notanorm/base.py [hunk 3]
```diff
@@ -5,47 +5,74 @@
...
-        return "DbRow(" + str(self.__dict__) + ")"
+        return "DbRow(" + super().__repr__() + ")"
 
     def __eq__(self, other):
         return self.__dict__ == other.__dict__
 
+    def __getattr__(self, key):
+        return self[key.lower()]
+
+    def __setattr__(self, key, val):
```
Comment (higleyc) [inline notanorm/base.py:47]: Is there a particular reason to allow modification of a row? Also per the other code, seems like it should be `key.lower()`

File: notanorm/mysql.py [hunk 4]
```diff
@@ -1,15 +1,37 @@
...
 # driver for mysql
 
 
 class MySqlDb(DbBase):
     placeholder = "%s"
-    retry_errors = (MySQLdb.OperationalError, )
-    integrity_errors = (MySQLdb.IntegrityError, )
+
+    @staticmethod
+    def translate_error(exp):
```
Comment (higleyc) [inline notanorm/mysql.py:19]: No change needed, just a thought: if it winds up being useful to see the full orig exception, we could instead have this reraise exceptions `from exp`

File: notanorm/mysql.py [hunk 5]
```diff
@@ -27,5 +49,99 @@ def _get_primary(self, table):
...
+            idxmap[idxinfo["key_name"]].append(idxinfo["column_name"])
+
+        indexes = []
+        for name, fds in idxmap.items():
+            indexes.append(DbIndex(fds, primary=(name == "PRIMARY")))
+
+        return DbTable(columns=tuple(cols), indexes=tuple(indexes))
+
+    def column_model(self, info):
+        if info.type == "int(11)":
```
Comment (higleyc) [inline notanorm/mysql.py:128]: what's this?

File: notanorm/sqlite.py [hunk 6]
```diff
@@ -1,23 +1,175 @@
...
+        if match:
+            typ = DbType.TEXT
+            fixed = match[1] == "character"
+            size = int(match[2])
+        else:
+            try:
+                typ = cls._type_map_inverse[info.type]
+            except KeyError:
+                typ = DbType.ANY
+        print("info: ", info, typ)
```
Comment (higleyc) [inline notanorm/sqlite.py:99]: rm

**Review by higleyc — APPROVED** [review]

File: notanorm/base.py (hunk 3 above)
Comment (earonesty) [reply notanorm/base.py:47]: well, we weren't blocking it before, so i people could have been stuffing extra values in db rows

**Pull request closed without merging**